### PR TITLE
Remove span pinning associated with string.Create

### PR DESCRIPTION
**Pull request by stephentoub**
This is now possible with C# 11.

### CODE REVIEW

File: src/libraries/Common/src/System/HexConverter.cs [hunk 1]
```diff
@@ -208,14 +208,10 @@ public static unsafe string ToString(ReadOnlySpan<byte> bytes, Casing casing = C
...
-            fixed (byte* bytesPtr = bytes)
-            {
-                return string.Create(bytes.Length * 2, (Ptr: (IntPtr)bytesPtr, bytes.Length, casing), static (chars, args) =>
-                {
-                    var ros = new ReadOnlySpan<byte>((byte*)args.Ptr, args.Length);
-                    EncodeToUtf16(ros, chars, args.casing);
-                });
-            }
+#pragma warning disable CS8500 // takes address of managed type
+            return string.Create(bytes.Length * 2, (RosPtr: (IntPtr)(&bytes), casing), static (chars, args) =>
```
Comment (MihaZupan) [inline src/libraries/Common/src/System/HexConverter.cs:212]: So this isn't actually taking the address of the underlying memory, just of the `ROS` struct from the stack? Neat

Comment (stephentoub) [reply src/libraries/Common/src/System/HexConverter.cs:212]: Correct

**Comment by jakobbotsch** [comment]
Keep in mind that this pattern introduces new address exposure, so it might be worth it to double check that it does not impact the preceding uses of the span in the same method significantly. If you want to be on the safe side you can create a copy of the span.

For example
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
private static unsafe string Foo(ReadOnlySpan<char> chars)
{
    for (int i = 0; i < chars.Length; i++)
    {
        if (chars[i] >= 128)
            return null;
    }

    //return string.Create(chars.Length, (IntPtr)(&chars), static (s, c) =>
    //{
    //    for (int j = 0; j < s.Length; j++)
    //        s[j] = (*(ReadOnlySpan<char>*)c)[j];
    //});
    fixed (char* charsPtr = chars)
    {
        return string.Create(chars.Length, (IntPtr)charsPtr, static (s, c) =>
        {
            for (int j = 0; j < s.Length; j++)
                s[j] = ((char*)c)[j];
        });
    }
}
```

Codegen for the uncommented vs commented versions of the the loop on linux-x64:
```diff
 G_M27558_IG03:
-       mov      eax, esi
-       cmp      word  ptr [rdi+2*rax], 128
-       jae      G_M27558_IG09
-       inc      esi
-       cmp      esi, ebx
+       mov      rsi, bword ptr [rbp-28H]
+       mov      eax, edi
+       cmp      word  ptr [rsi+2*rax], 128
+       jae      G_M27558_IG07
+       inc      edi
+       cmp      edi, dword ptr [rbp-20H]
        jl       SHORT G_M27558_IG03
```

(Fixing this would be the "partial lifetime promotion" item in #76931)

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.cs [hunk 2]
```diff
@@ -686,132 +686,78 @@ private static unsafe string JoinInternal(ReadOnlySpan<char> first, ReadOnlySpan
...
-                    (byte)(firstNeedsSeparator | secondNeedsSeparator << 1 | thirdNeedsSeparator << 2));
-
-                return string.Create(
-                    first.Length + second.Length + third.Length + fourth.Length + firstNeedsSeparator + secondNeedsSeparator + thirdNeedsSeparator,
-                    (IntPtr)(&payload),
-                    static (destination, statePtr) =>
+#pragma warning disable CS8500 // This takes the address of, gets the size of, or declares a pointer to a managed type
+            return string.Create(
+                first.Length + second.Length + third.Length + fourth.Length + firstNeedsSeparator + secondNeedsSeparator + thirdNeedsSeparator,
+                (FirstRos: (IntPtr)(&first), SecondRos: (IntPtr)(&second), ThirdRos: (IntPtr)(&third), FourthRos: (IntPtr)(&fourth), firstNeedsSeparator, secondNeedsSeparator, thirdNeedsSeparator),
```
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.cs:723]: Please don't reintroduce large tuples, it will regress size

**Comment by jakobbotsch** [comment]
> Ugh, it reintroduces bounds checking on the prior loop? That's not something we can fix?

Not easily. Address exposure turns off lots of basic analyses and optimizations for the exposed local and we're very coarse grained about it today.

**Comment by stephentoub** [comment]
@jakobbotsch, does it look ok now? Thanks.

**Review by jakobbotsch — APPROVED** [review]

**Pull request merged**